### PR TITLE
Templatize class factories.

### PR DESCRIPTION
**Pull request by anntzer**
This makes mpl_toolkits axes classes picklable (see test_pickle) by
generalizing the machinery of _picklable_subplot_class_constructor,
which would otherwise have had to be reimplemented for each class
factory.

## PR Checklist

- [ ] Has pytest style unit tests (and `pytest` passes).
- [ ] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [ ] New features are documented, with examples if plot related.
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [ ] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [ ] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by anntzer** [comment]
rebased, should be good to go.

**Comment by jklymak** [comment]
Since the song and dance was for cartopy do we know it works for them now?  @dopplershift @QuLogic

**Comment by dopplershift** [comment]
Examples seem to run fine (and there are no new test failures) with this PR + CartoPy master.

**Review by jklymak — APPROVED** [review]

File: lib/matplotlib/cbook/__init__.py [hunk 1]
```diff
@@ -2191,3 +2191,68 @@ def _unikey_or_keysym_to_mplkey(unikey, keysym):
...
+        subcls.__name__ = subcls.__qualname__ = fmt.format(base_class.__name__)
+        if attr_name is not None:
+            setattr(subcls, attr_name, base_class)
+        return subcls
+
+    return class_factory
+
+
+def _picklable_class_constructor(
+        base_cls, fmt, attr_name, _default_base_class, base_class):
```
Comment (QuLogic) [inline lib/matplotlib/cbook/__init__.py:2249]: Surely there's a better name to use than `base_cls` and `base_class`?

File: lib/matplotlib/axes/_subplots.py [hunk 2]
```diff
@@ -170,59 +159,10 @@ def _make_twin_axes(self, *args, **kwargs):
...
-    given axes_class (which is assumed to be a subclass of `.axes.Axes`).
-    This is perhaps a little bit roundabout to make a new class on
-    the fly like this, but it means that a new Subplot class does
-    not have to be created for every type of Axes.
-    """
-    if axes_class is None:
-        _api.warn_deprecated(
-            "3.3", message="Support for passing None to subplot_class_factory "
-            "is deprecated since %(since)s; explicitly pass the default Axes "
-            "class instead. This will become an error %(removal)s.")
```
Comment (QuLogic) [inline lib/matplotlib/axes/_subplots.py:198]: This is to be removed in 3.5, and seems to be the only place with default class, so all handling for  that can be dropped now?

File: lib/matplotlib/cbook/__init__.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/cbook/__init__.py:2249]: oops, the first one should have been mixin_class anyways...

File: lib/matplotlib/axes/_subplots.py (hunk 2 above)
Comment (anntzer) [reply lib/matplotlib/axes/_subplots.py:198]: yes

**Comment by anntzer** [comment]
rebased

File: lib/mpl_toolkits/axisartist/floating_axes.py [hunk 3]
```diff
@@ -351,12 +350,8 @@ def adjust_axes_lim(self):
...
 
 
-@functools.lru_cache(None)
-def floatingaxes_class_factory(axes_class):
-    return type("Floating %s" % axes_class.__name__,
-                (FloatingAxesBase, axes_class), {})
-
-
+floatingaxes_class_factory = cbook._make_class_factory(
+    FloatingAxesBase, "Floating {}")
```
Comment (QuLogic) [inline lib/mpl_toolkits/axisartist/floating_axes.py:354]: I guess this is an old typo, but surely this should not have a space in the name?

Comment (anntzer) [reply lib/mpl_toolkits/axisartist/floating_axes.py]: removed the space.

**Comment by QuLogic** [comment]
But I guess now that things have reasonable names, they're no longer linking in docs?

**Comment by anntzer** [comment]
Indeed, updated missing-references.

**Review by QuLogic — APPROVED** [review]

**Pull request merged**